[EASI-2417] Advice form recommendations

aterstriep opened the pull request:
# EASI-2417

[Figma](https://www.figma.com/file/iGatgTnlzTuqW9CQFYHDam/TRB-Intake-e2e?node-id=280%3A41427)

## Changes and Description

- Adds Recommendations section to Advice Letter form
- Advice Letter form unit tests
- Backend update to link form


<!-- Put a description here! -->

## How to test this change

- Follow instructions in #1937 for setting up an advice letter form for testing
- Test create, edit, and delete functionality on recommendations section of form

## PR Author Review Checklist

- [ ] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [ ] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [ ] Tested user-facing changes with voice-over and the [rotor menu](https://support.apple.com/guide/voiceover/with-the-voiceover-rotor-mchlp2719/mac)


## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

patrickseguraoddball commented on src/views/TechnicalAssistance/AdviceLetterForm/Recommendations.tsx:
```diff
@@ -1,75 +1,167 @@
+                {t('adviceLetterForm.noRecommendations')}
+              </Alert>
+            ) : (
+              <RecommendationsList
+                type="form"
+                recommendations={recommendations}
```
This could probably use some grid responsiveness.  On mobile it can get a bit scrunched

patrickseguraoddball commented on src/views/TechnicalAssistance/AdminHome/components/ReviewAdviceLetter/RecommendationLinks.tsx:
```diff
@@ -0,0 +1,46 @@
+              target="_blank"
+              rel="noopener noreferrer"
+              className="display-block margin-top-05"
+              // If EASi url, link to relative path
+              href={isExternal ? url.href : url.pathname}
+              // Display external link icon if not EASi url
+              {...(isExternal && { variant: 'external' })}
+            >
```
Nice accounting for EASI links!

patrickseguraoddball commented on src/views/TechnicalAssistance/AdminHome/components/ReviewAdviceLetter/RecommendationsList.tsx:
```diff
@@ -0,0 +1,120 @@
+  className?: string;
+};
+
+const CardTitle = ({ type, children, className }: CardTitleProps) => {
+  if (type === 'admin') {
+    return <h3 className={className}>{children}</h3>;
+  }
+  return <h4 className={className}>{children}</h4>;
+};
```
This might be a little overkill for something that could be achieved by a ternary operator

```
{type === 'admin' ? <h3 className={className}>{children}</h3> : <h4 className={className}>{children}</h4>
```

adamodd replied:
This is perfectly fine the way it is.

patrickseguraoddball commented on src/views/TechnicalAssistance/AdviceLetterForm/Recommendations.tsx:
```diff
@@ -1,75 +1,167 @@
+  });
+  const { reset } = formMethods;
 
+  const [remove, { loading }] = useMutation<
+    DeleteTRBRecommendation,
+    DeleteTRBRecommendationVariables
+  >(DeleteTrbRecommendationQuery, {
+    refetchQueries: [
       {
```
There needs to be some error handling here - notifications would behave similarly as outline in Figma

https://www.figma.com/file/iGatgTnlzTuqW9CQFYHDam/TRB-Intake-e2e?node-id=2240-107946&t=PFIUXk0T3E3AY82P-0

patrickseguraoddball commented on src/views/TechnicalAssistance/AdviceLetterForm/RecommendationsForm.tsx:
```diff
@@ -0,0 +1,259 @@
+          /** Format links to array of strings */
+          const links = (formData.links || []).map(({ link }) =>
+            // Format link
+            link.replace('www.', '').replace(/^https?:\/\//i, '')
```
This could maybe be an exported util, as it's used in a few places

patrickseguraoddball reviewed (changes requested):
Really nice work Ashley!  Just a few changes.  Also I think recently Zoe added the modal to warn before removing recommendation.  I think it should probably be added as a part of this ticket.

https://www.figma.com/file/iGatgTnlzTuqW9CQFYHDam/TRB-Intake-e2e?node-id=3882-122460&t=PFIUXk0T3E3AY82P-0

patrickseguraoddball commented on src/views/TechnicalAssistance/AdviceLetterForm/RecommendationsForm.tsx:
```diff
@@ -0,0 +1,262 @@
+        <Button
+          type="submit"
+          className="margin-top-6"
+          disabled={
+            watch('title').length === 0 ||
+            watch('recommendation').length === 0 ||
+            isSubmitting
+          }
```
I like how you handled this and is totally fine, just wanted to give an idea for something to scale with lots of fields.  Honestly not sure why RHF doesn't have something cleaner OOTB for this, not even sure if my recommendation works for all cases (conditionally required fields etc, but it works if all fields are required).  I want to look more into this.  No need to change this! 

```
disabled={
    Object.keys(dirtyFields).length !==
      Object.keys(defaultValues).length || isSubmitting
  }
```

patrickseguraoddball replied:
If it checks against the default/initial value, then if the value is reverted/erased, it would no longer be dirty.  I tested with with empty string and it seemed to work as expected.

patrickseguraoddball reviewed (changes requested):
Thanks for the changes Ashley.  

I noticed I can click through to Internal Review, bypassing required fields from previous steps

![Screenshot 2023-03-30 at 9 11 46 AM](https://user-images.githubusercontent.com/95709965/228846738-d8c556d3-a51c-4466-a997-40264627b695.png)

On mobile, the remove recommendation modal is bled through the screen size.  I think I introduced a fix to the modal/overflow hidden in a recent PR, but would be happy to sync on this.  Not a huge issue, but the fix is simple IIRC
![Screenshot 2023-03-30 at 9 12 54 AM](https://user-images.githubusercontent.com/95709965/228847057-5a515548-1908-4e2d-a862-288f74a243fd.png)

aterstriep commented:
> I noticed I can click through to Internal Review, bypassing required fields from previous steps

Good catch! This is fixed. When I updated the Next Steps required fields, I forgot to add validation for navigating through the step header. So it was only happening if you were on the Recommendations step and tried to click past Next Steps.

> On mobile, the remove recommendation modal is bled through the screen size. I think I introduced a fix to the modal/overflow hidden in a recent PR, but would be happy to sync on this. Not a huge issue, but the fix is simple IIRC

I think you're right about the recent PR and this is already fixed, but let me know if you're still seeing it.

patrickseguraoddball reviewed (approved):
Thanks Ashley!  The modal sizing thing is an issue but I think it's with the component in general and outside the scope of this PR, not need to tackle it now.
